openoa/utils/downloader.py

```diff
@@ -0,0 +1,425 @@
+
+    Monthly 10m height data is demonstrated here, as hourly data takes too long to download, 
+    but could be amended and other CDS datasets also used (e.g. CERRA for Europe)
+
+    Args:
+        lat(:obj:`float`): Latitude in WGS 84 spatial reference system (decimal degrees)
+        lon(:obj:`float`): Longitude in WGS 84 spatial reference system (decimal degrees)
+        save_pathname(:obj:`string`): The path where the downloaded reanalysis data will be saved
+        save_filename(:obj:`string`): The file name used to save the downloaded reanalysis data
+
+    Returns:
+        df(:obj:`dataframe`): A dataframe containing time series of the requested reanalysis variables
+        Saved NetCDF annual ERA5 files
+        Saved ERA5 csv file
+    """
+
+    logger.info("Please note access to ERA5 data requires registration")
+    logger.info("Please see: https://cds.climate.copernicus.eu/api-how-to")
+
+    # set up cds-api client
+    try:
+        c = cdsapi.Client()
+    except Exception as e:
+        logger.error('Failed to make connection to cds: '+ str(e))
+        logger.error('Please see: https://cds.climate.copernicus.eu/api-how-to')
+        raise NameError(e)
+
+    # create save_pathname if it does not exist
+    if not os.path.exists(save_pathname):
+        os.makedirs(save_pathname)
+
+    # downloads all years from today back to the year 2000
+    now = datetime.datetime.now()
+    years = list(range(2000,now.year+1,1))
+
+    # get the data for the closest 9 nodes to the coordinates
+    node_spacing = 0.250500001*1
+
+    # download the data
+    for year in years:
+
+        outfile = save_pathname+save_filename+"_"+str(year)+".nc"
+
+        if year == now.year:
+            months = list(range(1,now.month,1))
+        else:
+            months = list(range(1,12+1,1))
+
+        # See: https://cds.climate.copernicus.eu/cdsapp#!/dataset/reanalysis-era5-single-levels-monthly-means?tab=form
+        # for formulating other requests from cds
+        if not os.path.exists(outfile) or year==now.year:
+
+            logger.info("Downloading ERA5 :"+outfile)
+
+            try:
+                c.retrieve(
+                    "reanalysis-era5-single-levels-monthly-means",
+                    {
+                        "product_type": "reanalysis",
+                        "format": "netcdf",
+                        "variable": [
+                            "10m_wind_speed", "2m_temperature", "surface_pressure",
+                        ],
+                        "year": year,
+                        "month": months,
+                        "product_type": "monthly_averaged_reanalysis",
+                        "time": [
+                            "00:00"
+                        ],
+                        "area": [
+                            lat+node_spacing, lon-node_spacing, 
+                            lat-node_spacing, lon+node_spacing,
+                        ],
+                    },
```

Comment:
It might be better to define this as a base dictionary above and update it in each loop so that it's a little bit easier to follow along, and so that the dictionary isn't needing to be redefined for common variables in every iteration.

```python
cds_request = {
    "product_type": "reanalysis",  # example of static variable
    "year": 0  # example of dynamic variable
}
for year in year:
    # other stuff
    base_request.update("year", year)
    base_request.update("months", months)
    c.retrieve("reanalysis-era5-single-levels-monthly-means", cds_request, outfile)
```

Reply:
Done. Yep, I like this. 

I'm tempted to make it so that we have a generic `get_cds` function that then has the `cds_dataset` and `cds_request` as inputs, but I don't want to over complicate things for now, as there are a lot of datasets on cds and would need to check carefully something generic would work well with other datasets.

Reply:
Totally agreed, I don't think we need to get into the weeds here until we know we'll need more ERA5 data access!